rust/parquet/src/arrow/arrow_writer.rs

```diff
@@ -175,15 +175,61 @@ fn write_leaves(
             }
             Ok(())
         }
+        ArrowDataType::Dictionary(k, v) => {
+            // Materialize the packed dictionary and let the writer repack it
+            let any_array = array.as_any();
+            let (k2, v2) = match &**k {
+                ArrowDataType::Int32 => {
+                    let typed_array = any_array
+                        .downcast_ref::<arrow_array::Int32DictionaryArray>()
+                        .expect("Unable to get dictionary array");
+
+                    (typed_array.keys(), typed_array.values())
+                }
+                o => unimplemented!("Unknown key type {:?}", o),
+            };
+
+            let k3 = k2;
+            let v3 = v2
+                .as_any()
+                .downcast_ref::<arrow_array::StringArray>()
+                .unwrap();
+
+            // TODO: This removes NULL values; what _should_ be done?
+            // FIXME: Don't use `as`
+            let materialized: Vec<_> = k3
+                .flatten()
+                .map(|k| v3.value(k as usize))
+                .map(ByteArray::from)
+                .collect();
+            //
```

Comment:
It seems like it is ignored in the branch this PR opened against:

```
/// Get the underlying numeric array slice, skipping any null values.
/// If there are no null values, it might be quicker to get the slice directly instead of
/// calling this function.
fn get_numeric_array_slice<T, A>(array: &arrow_array::PrimitiveArray<A>) -> Vec<T::T>
where
    T: DataType,
    A: arrow::datatypes::ArrowNumericType,
    T::T: From<A::Native>,
{
    let mut values = Vec::with_capacity(array.len() - array.null_count());
    for i in 0..array.len() {
        if array.is_valid(i) {
            values.push(array.value(i).into())
        }
    }
    values
}
```

Thou other implementations are appending null values. We need some changes to adapt in Rust impl., which means ByteArray should have `From` implementation for `Vec<Option<u8>>` and `Option<&'a str>` then the aforementioned method should be adapted to nulls.

Related impl. in Cxx: https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/path_internal.cc#L51-L73

Reply:
@vertexclique I [updated the roundtrip dictionary test to include some `None` values](https://github.com/apache/arrow/pull/8402/commits/5fc3543ebbbf9c1a930a0e0383750418e49e70df), and it passes, so I think this code is fine-- it seems that the `None` values are handled by the `definition` levels, so we don't need to handle them here. Do I have that right or am I still missing something?

Reply:
Only seeing this now.

Yes, @carols10cents is correct. What makes Arrow > Parquet challenging is that we have to take an array like `<1, 2, null, 4, null, 6>` and convert it to `<1, 2, 4, 6>` then have definitions `<1, 1, 0, 1, 0, 1>`.
It's super-trivial for primitives, but once you start nesting; it becomes difficult to even reason about what happens on some exotic combinations, especially with nested lists.